### PR TITLE
impl(generator): add Start and Await overloads to clients

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #7658 

Adds the "Start" and "Await" overloads for each method_signature defined for the rpc to the Client class.

PR is split into multiple commits to separate changes, tests, and generated code.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370)

### CODE REVIEW

File: generator/internal/client_generator.cc [hunk 1]
```diff
@@ -253,7 +264,22 @@ R"""(  std::unique_ptr<::google::cloud::AsyncStreamingReadWriteRpc<
...
     "  future<StatusOr<$longrunning_deduced_response_type$>>\n"},
-   {"  $method_name$($request_type$ const& request, Options opts = {});\n"}
+   {"  $method_name$($request_type$ const& request, Options opts = {});\n"},
+   {"\n"},
+                 // clang-format on
+                 {IsResponseTypeEmpty,
+                  // clang-format off
+    "  Status\n",
+    "  StatusOr<$longrunning_operation_type$>\n"},
+   {"  $method_name$(google::cloud::ExperimentalTag, "
```
Comment (dbolduc) [inline generator/internal/client_generator.cc:274]: same nit about: `google::cloud::`

File: generator/integration_tests/golden/v1/golden_kitchen_sink_client.h [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #define GOOGLE_CLOUD_CPP_GENERATOR_INTEGRATION_TESTS_GOLDEN_V1_GOLDEN_KITCHEN_SINK_CLIENT_H
 
 #include "generator/integration_tests/golden/v1/golden_kitchen_sink_connection.h"
+#include "google/cloud/experimental_tag.h"
```
Comment (dbolduc) [inline generator/integration_tests/golden/v1/golden_kitchen_sink_client.h:23]: unused?

File: generator/integration_tests/golden/v1/golden_rest_only_client.h [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #define GOOGLE_CLOUD_CPP_GENERATOR_INTEGRATION_TESTS_GOLDEN_V1_GOLDEN_REST_ONLY_CLIENT_H
 
 #include "generator/integration_tests/golden/v1/golden_rest_only_rest_connection.h"
+#include "google/cloud/experimental_tag.h"
```
Comment (dbolduc) [inline generator/integration_tests/golden/v1/golden_rest_only_client.h:23]: unused?

File: generator/integration_tests/golden/v1/golden_thing_admin_client.cc [hunk 4]
```diff
@@ -57,12 +57,33 @@ GoldenThingAdminClient::CreateDatabase(std::string const& parent, std::string co
   return connection_->CreateDatabase(request);
 }
 
+StatusOr<google::longrunning::Operation>
+GoldenThingAdminClient::CreateDatabase(ExperimentalTag, NoAwaitTag, std::string const& parent, std::string const& create_statement, Options opts) {
```
Comment (dbolduc) [inline generator/integration_tests/golden/v1/golden_thing_admin_client.cc:61]: silly nit: should we name these tags and `std::move` them into the connection call?

File: generator/integration_tests/golden/v1/golden_thing_admin_client.h [hunk 5]
```diff
@@ -405,6 +419,12 @@ class GoldenThingAdminClient {
   future<StatusOr<google::test::admin::database::v1::UpdateDatabaseDdlMetadata>>
   UpdateDatabaseDdl(google::test::admin::database::v1::UpdateDatabaseDdlRequest const& request, Options opts = {});
 
+  StatusOr<google::longrunning::Operation>
```
Comment (dbolduc) [inline generator/integration_tests/golden/v1/golden_thing_admin_client.h:422]: We need to document the client methods.

And the documentation should probably mention that the `NoAwaitTag` overload is for advanced use cases.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370)* status: 0 of 705 files reviewed, 5 unresolved discussions (waiting on @dbolduc)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370#gh-1653640937:-O0K5C8mBqu0PJONABDB:b8fyg9p) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/14370#discussion_r1653640937) to the right spot — sorry!)*

same nit about: `google::cloud::`
</blockquote></details>

Thought I fixed that, must have lost it in a merge/rebase.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370#gh-1653656125:-O0K5ITvBIWbZeojWg1b:b-896fix) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/14370#discussion_r1653656125) to the right spot — sorry!)*

unused?
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370#gh-1653656168:-O0K5IrACRWWUXRjYnqv:b-896fix) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/14370#discussion_r1653656168) to the right spot — sorry!)*

unused?
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370#gh-1653657917:-O0K5JR66Mhdh5k89-tz:bbn7khp) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/14370#discussion_r1653657917) to the right spot — sorry!)*

silly nit: should we name these tags and `std::move` them into the connection call?
</blockquote></details>

Possibly, it will require changes to the connection generation as well. If we decide we want this, I'll do it in a separate PR.

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14370#gh-1653658761:-O0K5SFg4VqEM9f-VuTS:b6b4epl) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/14370#discussion_r1653658761) to the right spot — sorry!)*

We need to document the client methods.

And the documentation should probably mention that the `NoAwaitTag` overload is for advanced use cases.
</blockquote></details>

The documentation PR is next. I need to document the client functions, add the EXPECT_CALL comments on the preexisting connection functions, as well as the blurb in the CHANGELOG detailing the change in testing behavior.

___

**Review by dbolduc — APPROVED** [review]

**Pull request merged**